src/probability_theory/density.lean

```diff
@@ -0,0 +1,348 @@
+
+## Main results
+
+* `measure_theory.pdf.integral_mul_eq_integral'` : Law of the unconscious statistician, i.e.
+  if a random variable `X : α → E` has pdf `f`, then `𝔼(g(X)) = ∫ x, g x * f x dx` for
+  all measurable `g : E → ℝ`.
+* `measure_theory.pdf.integral_mul_eq_integral` : A real-valued random variable `X` with
+  pdf `f` has expectation `∫ x, x * f x dx`.
+* `measure_theory.pdf.uniform.integral_eq` : If `X` follows the uniform distribution with
+  its pdf having support `s`, then `X` has expectation `(λ s)⁻¹ * ∫ x in s, x dx` where `λ`
+  is the Lebesgue measure.
+
+## TODOs
+
+Ultimately, we would also like to define characteristic functions to describe distributions as
+it exists for all random variables. However, to define this, we will need Fourier transforms
+which we currently do not have.
+-/
+
+noncomputable theory
+open_locale classical measure_theory nnreal ennreal
+
+namespace measure_theory
+
+open topological_space measure
+
+variables {α E : Type*} [normed_group E] [measurable_space E] [second_countable_topology E]
+  [normed_space ℝ E] [complete_space E] [borel_space E]
+
+/-- A random variable `X : α → E` is said to `has_pdf` with respect to the measure `ℙ` on `α` and
+`μ` on `E` if there exists a measurable function `f` such that the push-forward measure of `ℙ`
+along `X` equals `μ.with_density f`. -/
+class has_pdf {m : measurable_space α} (X : α → E)
+  (ℙ : measure α) (μ : measure E . volume_tac) : Prop :=
+(pdf' : ∃ (f : E → ℝ≥0∞), measurable f ∧ map X ℙ = μ.with_density f)
+
+/-- If `X` is a random variable that `has_pdf X ℙ μ`, then `pdf X` is the measurable function `f`
+such that the push-forward measure of `ℙ` along `X` equals `μ.with_density f`. -/
+def pdf {m : measurable_space α} (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac) :=
+if hX : has_pdf X ℙ μ then classical.some hX.pdf' else 0
+
+@[measurability]
+lemma measurable_pdf {m : measurable_space α}
+  (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac) :
+  measurable (pdf X ℙ μ) :=
+begin
+  by_cases hX : has_pdf X ℙ μ,
+  { rw [pdf, dif_pos hX],
+    exact (classical.some_spec hX.pdf').1 },
+  { rw [pdf, dif_neg hX],
+    exact measurable_zero }
+end
+
+lemma pdf_spec {m : measurable_space α}
+  (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac) [hX : has_pdf X ℙ μ] :
+  measure.map X ℙ = μ.with_density (pdf X ℙ μ) :=
+begin
+  rw [pdf, dif_pos hX],
+  exact (classical.some_spec hX.pdf').2
+end
+
+lemma pdf_spec' {m : measurable_space α}
+  (X : α → E) (ℙ : measure α) (μ : measure E . volume_tac) [hX : has_pdf X ℙ μ]
+  {s : set E} (hs : measurable_set s) :
+  measure.map X ℙ s = ∫⁻ x in s, pdf X ℙ μ x ∂μ :=
+by rw [← with_density_apply _ hs, pdf_spec X ℙ μ]
+
+namespace pdf
+
+variables {m : measurable_space α} {ℙ : measure α} {μ : measure E}
+
+section
+
+lemma lintegral_eq_measure_univ {X : α → E} [has_pdf X ℙ μ] (hX : measurable X) :
+  ∫⁻ x, pdf X ℙ μ x ∂μ = ℙ set.univ :=
```

Comment:
Does it make sense to allow dot notation such that this can instead be spelt
```suggestion
lemma lintegral_eq_measure_univ {X : α → E} [ℙ.has_pdf X μ] (hX : measurable X) :
  ∫⁻ x, ℙ.pdf X μ x ∂μ = ℙ set.univ :=
```

Reply:
Sure, I will put it in the `measure` namespace. I'm not sure if I want to change `has_pdf X ℙ μ` to `ℙ.has_pdf X μ` in this file however since I read the statement as "X has pdf wrt μ". This is also how `quasi_measure_preserving` is done also (defined in the measure namespace but does not use dot notation).